Feature/user input validation

Marcos-Peter opened the pull request.

Marcos-Peter committed:
Finishing the user input validation and restructuring some files

Marcos-Peter committed:
Fixing some dependency versions

Marcos-Peter committed:
Add the password encrypt using bcrypt

alanraso commented on .editorconfig:
```diff
@@ -0,0 +1,12 @@
+# EditorConfig is awesome: https://EditorConfig.org
+
+# top-most EditorConfig file
+root = true
```
@Marcos-Peter 
Dá uma olhada no posicionamento das branches e tenta fazer com que essa aqui tenha apenas os commits relativos à feature que vc quer que seja revista, pq ela tá com bastante código da anterior.

alanraso commented on src/index.ts:
```diff
@@ -1,22 +1,20 @@
-    console.table(users);
-  })
-  .then(apolloServerRun)
+AppDataSource.initialize()
+  .then(async () => console.log(`Connected to database.`))
   .catch((error) => console.log(error));
+
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+});
+
+server.listen({ port }).then(({ url }) => {
+  console.log(`[SERVER] - Server running at ${url}`);
+});
```
Aqui ele vai conectar com o banco e em paralelo vai iniciar o servidor. Porém, se a conexão com o banco falhar, o servidor vai continuar rodando e quase 100% dele não vai funcionar. Não é melhor interromper ou nem startar o servidor caso a conexão com o banco falhe?

Marcos-Peter replied:
Com certeza! Acabei não me atentando a esse caso, mas já foi corrigido. Muito obrigado pela correção!

alanraso commented on src/resolvers/resolver.ts:
```diff
@@ -1,27 +1,52 @@
+const userRepo = AppDataSource.getRepository(User);
+
+async function validateInput(userData: UserInput) {
+  const passwordRegex = new RegExp(/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])[0-9a-zA-Z]{8,16}$/);
```
Esse tipo de constante não precisa ser redefinida a cada execução da função, ou seja, ela pode ficar na fora dela, certo?

Marcos-Peter replied:
Correto! Corrigido, obrigado!

alanraso commented on src/resolvers/resolver.ts:
```diff
@@ -1,27 +1,52 @@
+
+async function validateInput(userData: UserInput) {
+  const passwordRegex = new RegExp(/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])[0-9a-zA-Z]{8,16}$/);
+  if (passwordRegex.test(userData.password) == false) {
```
```suggestion
  if (!passwordRegex.test(userData.password)) {
```

É legal evitar `==` e usar sempre `===`. Ou nesse caso, como simplificação, qualquer coisa diferente de true ele deve jogar o erro.

alanraso commented on src/resolvers/resolver.ts:
```diff
@@ -1,27 +1,52 @@
+    throw new Error('Email address already in use.');
+  }
+
+  if (!userData.name) {
+    throw new Error('Name is required.');
+  } else if (userData.name.length < 3) {
+    throw new Error('Name must have at least 3 characters.');
+  } else if (!userData.name.trim()) {
+    throw new Error('Name cannot be empty.');
+  }
+
+  if (!userData.birthdate) {
+    throw new Error('Birthdate is required.');
+  }
+  if (!new Date(userData.birthdate).getTime()) {
+    throw new Error('Birthdate must be a valid date.');
+  }
```
O GraphQL já garante a obrigatoriedade dos campos, não é necessário fazer isso manualmente.

Marcos-Peter committed:
Update password encripter class

company conventions

Marcos-Peter committed:
Fixing the class import

Marcos-Peter committed:
Merge pull request #15 from indigotech/feature/encrypt-password

Feature/encrypt password

Marcos-Peter committed:
Restructuring the server connection

Ensure that the server only starts if the connection to the database is successful

Marcos-Peter reviewed (commented):
Alterações realizadas

Marcos-Peter committed:
Update resolver.ts

extracted regex constant

alanraso commented on src/index.ts:
```diff
@@ -1,22 +1,37 @@
+  resolvers,
+});
+
+const startServer = server.listen({ port }).then(({ url }) => {
+  console.log(`[SERVER] - Server running at ${url}`);
+});
+
+export async function runServer(isMain = false) {
+  await AppDataSource.initialize()
+    .then(() => {
+      console.log(`[SERVER] - Connected to database.`);
+      startServer;
```
Aqui ele vai rodar o listen na própria criação da variável `startServer` e vai acabar não fazendo nada aqui. Recomendo usar a sintaxe de async/await para as 2 operações, fica mais fácil de organizar:

```ts
async function setupDatabase() {
  await AppDataSource.initialize();
  console.info(`[SERVER] - Connected to database.`);
}

async function setupServer() {
  const { url } = awai server.listen({ port });
  console.info(`[SERVER] - Server running at ${url}`);
}

async function setup() {
  await setupDatabase();
  await setupServer();
}
```

Obs:
- O `console.info` é mais apropriado pra esses logs porque eles são informativos e não usados para debuggar algo, como costuma ser o `console.log`
- Daria pra usar try-catch pra pegar o erro de conexão com o banco por exemplo, porém se vc deixar ele esoutrar, ele interrompe a execução e já loga o erro, então creio que não seja necessário.

Marcos-Peter committed:
Removing unnecessary validations
